### PR TITLE
Add ability to create a TrustContext from PEM/DER certs in memory

### PR DESCRIPTION
**Pull request by jared2501**

### CODE REVIEW

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java [hunk 1]
```diff
@@ -80,12 +83,43 @@ private KeyStores() {
...
+    static KeyStore createTrustStoreFromCertificates(Map<String, byte[]> certificatesByAlias) {
         try {
             KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
             keyStore.load(null, null);
 
-            for (File currFile : getFilesForPath(path)) {
-                keyStore.setCertificateEntry(currFile.getName(), readX509Certificate(currFile.toPath()));
+            for (Map.Entry<String, byte[]> entry : certificatesByAlias.entrySet()) {
+                keyStore.setCertificateEntry(
+                        entry.getKey(), readX509Certificate(new ByteArrayInputStream(entry.getValue())));
```
Comment (markelliot) [inline ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java]: you don't close the BAIS

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java [hunk 2]
```diff
@@ -80,12 +83,43 @@ private KeyStores() {
...
+    /**
+     * Returns a {@link KeyStore} created by loading the given certificates.
+     *
+     * @param certificatesByAlias
+     *        a map of X.509 certificate in PEM or DER format by the alias to load the certificate as.
+     */
+    static KeyStore createTrustStoreFromCertificates(Map<String, byte[]> certificatesByAlias) {
         try {
             KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
             keyStore.load(null, null);
```
Comment (markelliot) [inline ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java]: these two lines should mirror the function above, which catch the exception near here and provide more granular per-certificate exception handling below

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java [hunk 3]
```diff
@@ -80,12 +83,43 @@ private KeyStores() {
...
+        return keyStore;
+    }
+
+    /**
+     * Returns a {@link KeyStore} created by loading the given certificates.
+     *
+     * @param certificatesByAlias
+     *        a map of X.509 certificate in PEM or DER format by the alias to load the certificate as.
+     */
+    static KeyStore createTrustStoreFromCertificates(Map<String, byte[]> certificatesByAlias) {
```
Comment (markelliot) [inline ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java]: I think `byte[]` is probably a weird way to carry the certificate bytes, maybe we should make an alias object so it's obvious what it is?

**Review by markelliot — CHANGES_REQUESTED** [review]
not a huge fan of the code duplication, but probably needs a slightly bigger refactor to eliminate it, and probably isn't worth the refactor. (ie could have a `CertificateCreationStrategy<T>` and then create fixed instances of it to pass to the other methods, meh)

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java (hunk 3 above)
Comment (jared2501) [reply ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java]: Sweet - how does `PemX509Certificate` sound as alias?

**Comment by markelliot** [comment]
Works for me.

**Comment by uschi2000** [comment]
wow, this github diff is quite terrible :(

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/SslSocketFactories.java [hunk 4]
```diff
@@ -46,6 +49,11 @@ public static SSLSocketFactory createSslSocketFactory(SslConfiguration config) {
         return sslContext.getSocketFactory();
     }
 
+    public static SSLSocketFactory createSslSocketFactory(Map<String, byte[]> trustCertificatesByAlias) {
```
Comment (uschi2000) [inline ssl-config/src/main/java/com/palantir/remoting1/config/ssl/SslSocketFactories.java]: needs docs

Comment (uschi2000) [reply ssl-config/src/main/java/com/palantir/remoting1/config/ssl/SslSocketFactories.java]: you'll have to explain somewhere what exactly you mean by "trustCertificatesByAlias"

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java (hunk 1 above)
Comment (jared2501) [reply ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java]: technically, the close method on the BAIS does nothing:

```
public void close() throws IOException {
    }
```

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java (hunk 3 above)
Comment (jared2501) [reply ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java]: done

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java (hunk 2 above)
Comment (jared2501) [reply ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java]: done

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java (hunk 3 above)
Comment (jared2501) [reply ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java]: done

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java (hunk 1 above)
Comment (jared2501) [reply ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java]: done

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/SslSocketFactories.java (hunk 4 above)
Comment (jared2501) [reply ssl-config/src/main/java/com/palantir/remoting1/config/ssl/SslSocketFactories.java]: done

**Comment by jared2501** [comment]
addressed comments

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java [hunk 5]
```diff
@@ -80,18 +82,57 @@ private KeyStores() {
...
      */
     static KeyStore createTrustStoreFromCertificates(Path path) {
+        KeyStore keyStore;
         try {
-            KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+            keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
             keyStore.load(null, null);
+        } catch (IOException | GeneralSecurityException e) {
+            throw Throwables.propagate(e);
+        }
```
Comment (markelliot) [inline ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java]: pull out this block of code into `private static KeyStore createKeyStore()`

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/SslSocketFactories.java [hunk 6]
```diff
@@ -74,6 +88,24 @@ public static SSLContext createSslContext(SslConfiguration config) {
...
+    public static SSLContext createSslContext(Map<String, PemX509Certificate> trustCertificatesByAlias) {
+        TrustManager[] trustManagers = createTrustManagers(trustCertificatesByAlias);
+
+        try {
+            SSLContext sslContext = SSLContext.getInstance("TLS");
+            sslContext.init(new KeyManager[]{}, trustManagers, null);
+            return sslContext;
+        } catch (GeneralSecurityException e) {
+            throw Throwables.propagate(e);
+        }
```
Comment (markelliot) [inline ssl-config/src/main/java/com/palantir/remoting1/config/ssl/SslSocketFactories.java:102]: guessing this block of code also repeated: `private static SSLContext createSslContextFromTrustManagers(TrustManager[] trustManagers)`?

**Review by markelliot — CHANGES_REQUESTED** [review]

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java (hunk 5 above)
Comment (jared2501) [reply ssl-config/src/main/java/com/palantir/remoting1/config/ssl/KeyStores.java]: done

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/SslSocketFactories.java (hunk 6 above)
Comment (jared2501) [reply ssl-config/src/main/java/com/palantir/remoting1/config/ssl/SslSocketFactories.java:102]: done

**Review by jared2501 — COMMENTED** [review]
done review

**Review by markelliot — COMMENTED** [review]
lgtm, Rob?

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/PemX509Certificate.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import org.immutables.value.Value;
+
+@Value.Immutable
+@Value.Style(visibility = Value.Style.ImplementationVisibility.PACKAGE, jdkOnly = true)
+@JsonSerialize(as = ImmutablePemX509Certificate.class)
+@JsonDeserialize(as = ImmutablePemX509Certificate.class)
+public abstract class PemX509Certificate {
+
+    /** A X509.1 certificate in PEM format encoded in UTF-8. */
+    public abstract String pemCertificate();
```
Comment (uschi2000) [inline ssl-config/src/main/java/com/palantir/remoting1/config/ssl/PemX509Certificate.java:34]: what's the deal here with those PEM headers, something like `---PUBLIC KEY---`?

Comment (jared2501) [reply ssl-config/src/main/java/com/palantir/remoting1/config/ssl/PemX509Certificate.java:34]: they can be anything - `PUBLIC KEY`, `CERTIFICATE`, etc.. I don't think it actually matters. Can write test if you wish?

**Comment by uschi2000** [comment]
just add a comment, maybe?

**Comment by uschi2000** [comment]
or a test too if it's simple enough.

File: ssl-config/src/main/java/com/palantir/remoting1/config/ssl/PemX509Certificate.java (hunk 7 above)
Comment (nmiyake) [reply ssl-config/src/main/java/com/palantir/remoting1/config/ssl/PemX509Certificate.java:34]: Haven't verified behavior myself, but sounds like content after `BEGIN` and `END` may matter -- see https://tools.ietf.org/html/rfc7468#page-11 and http://stackoverflow.com/a/5356351. May be worth verifying/investigating.

Comment (jared2501) [reply ssl-config/src/main/java/com/palantir/remoting1/config/ssl/PemX509Certificate.java:34]: done - added this comment:

```
    /**
     * A X509.1 certificate or certificate chain in PEM format encoded in UTF-8.
     * <p>
     * The certificates must be delimited by the the begin and end {@code CERTIFICATE} markers.
     */

```

**Review by uschi2000 — APPROVED** [review]

**Comment by uschi2000** [comment]
@markelliot I think you'll have to "approve" explicitly

**Review by markelliot — APPROVED** [review]

**Pull request merged**